### PR TITLE
[fix]: [CDS-79062]: Index creation asynchronously

### PR DESCRIPTION
**Pull request by archit-harness**
## Describe your changes

Changes - Index creation asynchronously

Impact - Index creation will happen asynchronously without impacting service startup time and restarts

Testing -

a) Tested creating new index in the collection

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52799)

### CODE REVIEW

**Comment by archit-harness** [comment]
trigger uts

File: 960-persistence/src/main/java/io/harness/mongo/index/IndexManagerSession.java [hunk 1]
```diff
@@ -383,15 +397,29 @@ public void create(IndexCreator indexCreator) {
...
             if (exception.getErrorCode() == 24) {
               Morpheus.sleep(Duration.ofSeconds(1));
             } else {
+              log.warn("Index creation failed for {} {}", indexCreator.getOptions().toString(),
+                  indexCreator.getKeys().toString());
               throw exception;
             }
+
+          } catch (TimeoutException | InterruptedException | ExecutionException e) {
+            log.warn("Index creation failed for {} {}", indexCreator.getOptions().toString(),
```
Comment (hextechpal) [inline 960-persistence/src/main/java/io/harness/mongo/index/IndexManagerSession.java:420]: I think this is misleading,timeout exception just indicates that the future did not complete in the specified time which is expected for larger collections

I think we should change this to 
`
catch (TimeoutException e) {
            log.warn("")
}catch(InterruptedException | ExecutionException e){
            //This indicates that the creation truly failed
            log.error("")
}
`

File: 960-persistence/src/main/java/io/harness/mongo/index/IndexManagerSession.java [hunk 2]
```diff
@@ -455,19 +489,54 @@ public boolean rebuildIndex(
...
+      if (tempTRITimeStamp + waitFor.toMillis() <= currentTimeMillis()) {
+        rebuildIndex = true;
+      }
+    }
+
+    try {
+      Future<?> future = createIndexExecutorService.submit(
+          () -> handlingTRIAndCollidedIndexes(collectionSession, triIndex, tempCreator, indexCreator, waitFor));
+      if (completeCreateFuture) {
+        future.get(4, TimeUnit.HOURS);
```
Comment (hextechpal) [inline 960-persistence/src/main/java/io/harness/mongo/index/IndexManagerSession.java:514]: This is a blocking operation for 4 hours, is this correct?

**Review by hextechpal — CHANGES_REQUESTED** [review]

File: 960-persistence/src/main/java/io/harness/mongo/index/IndexManagerSession.java [hunk 3]
```diff
@@ -97,6 +104,9 @@ public class IndexManagerSession {
   public static final Duration SOAKING_PERIOD = ofDays(1);
   // We do not want to drop temporary index before we created the new one. Make the soaking period smaller.
   public static final Duration REBUILD_SOAKING_PERIOD = SOAKING_PERIOD.minus(ofHours(1));
+
+  private ExecutorService createIndexExecutorService = ThreadPool.create(
```
Comment (abhinav-harness) [inline 960-persistence/src/main/java/io/harness/mongo/index/IndexManagerSession.java:108]: async index creation can cause newer queries to be slowed down

**Review by archit-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/52799)* status: 3 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-persistence/src/main/java/io/harness/mongo/index/IndexManagerSession.java` line 108 at r2](https://reviewable.io/reviews/harness/harness-core/52799#gh-1324004112:-NeC5wnO-ueol_7opvAN:b-1aq217) ([raw file](https://github.com/harness/harness-core/blob/a784652ff37f0c83ccf6e339c80a19fdda39eb54/960-persistence/src/main/java/io/harness/mongo/index/IndexManagerSession.java#L108)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

async index creation can cause newer queries to be slowed down
</blockquote></details>

Yes, but if indexes takes a lot of time, we expect them to be created manually rather than through AUTO mode, it anyway happen if index creation doesn't get completed within few mins today as service restart happens
___
*[`960-persistence/src/main/java/io/harness/mongo/index/IndexManagerSession.java` line 514 at r2](https://reviewable.io/reviews/harness/harness-core/52799#gh-1324003581:-NeC5uCh6FfBqzKMSwEA:bvv3h5e) ([raw file](https://github.com/harness/harness-core/blob/a784652ff37f0c83ccf6e339c80a19fdda39eb54/960-persistence/src/main/java/io/harness/mongo/index/IndexManagerSession.java#L514)):*
<details><summary><i>Previously, hextechpal (Prashant Pal) wrote…</i></summary><blockquote>

This is a blocking operation for 4 hours, is this correct?
</blockquote></details>

yes, for large collection kept higher value

**Pull request closed without merging**